export.go:123 Crashes with segmentation fault when running prometheus one shot.

Here is a summary of the error:

`
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x47f1e2]

goroutine 1 [running]:
sync.(*WaitGroup).state(...)
        /usr/lib/golang/src/sync/waitgroup.go:40
sync.(*WaitGroup).Add(0x0?, 0x0?)
        /usr/lib/golang/src/sync/waitgroup.go:63 +0x22
github.com/google/mtail/internal/exporter.New({0xb55cd8?, 0xc00043c380}, 0x0, 0xc0001a8fc0, {0xc00060fb20, 0x0, 0x7fc447000000?})
        ./mtail/internal/exporter/export.go:123 +0x7a5
main.main()
        ./mtail/cmd/mtail/main.go:251 +0x1e70
`
And the command that generated it:

`./bin/mtail -progs my_rules --one_shot --one_shot_format prometheus --logs test.log`

my_rules contains a file called testrule.mtail with contents:

The test.log file just contains 5 lines:

`
a

a

a

a

a
`

`
counter lines_total
/$/ {
  lines_total++
}
`

It appears that in cmd/mtail/main.go:251

nil is passed as the second parameter into the exporter.New(). This eventually gets passed all the way down to 
export.go:123 where the waitGroug, which is nil, attempts to call wg.Add(1) and crashes.

One way to possibly fix this is to wrap the portions of code that access the wg variable in a 

` if wg != nil `

Statement.

--- a/cmd/mtail/main.go
+++ b/cmd/mtail/main.go
@@ -11,6 +11,7 @@ import (
 	"os/signal"
 	"runtime"
 	"strings"
+	"sync"
 	"syscall"
 	"time"
 
@@ -248,19 +249,23 @@ func main() {
 	if *oneShot {
 		switch *oneShotFormat {
 		case "prometheus":
-			e, err := exporter.New(ctx, nil, store, eOpts...)
+			var wg sync.WaitGroup
+			e, err := exporter.New(ctx, &wg, store, eOpts...)
 			if err != nil {
 				glog.Error(err)
 				cancel()
+				wg.Wait()
 				os.Exit(1) //nolint:gocritic // false positive
 			}
 			err = e.Write(os.Stdout)
 			if err != nil {
 				glog.Error(err)
 				cancel()
+				wg.Wait()
 				os.Exit(1) //nolint:gocritic // false positive
 			}
 			cancel()
+			wg.Wait()
 			os.Exit(0) //nolint:gocritic // false positive
 		case "json":
 			err = store.WriteMetrics(os.Stdout)

--- a/internal/exporter/export.go
+++ b/internal/exporter/export.go
@@ -88,6 +88,9 @@ func New(ctx context.Context, wg *sync.WaitGroup, store *metrics.Store, options 
 	if store == nil {
 		return nil, errors.New("exporter needs a Store")
 	}
+	if wg == nil {
+		return nil, errors.New("exporter needs a WaitGroup")
+	}
 	e := &Exporter{
 		ctx:      ctx,
 		store:    store,

--- a/internal/exporter/export.go
+++ b/internal/exporter/export.go
@@ -83,13 +83,18 @@ func DisableExport() Option {
 	}
 }
 
+var (
+	ErrNeedsStore     = errors.New("exporter needs a Store")
+	ErrNeedsWaitgroup = errors.New("exporter needs a WaitGroup")
+)
+
 // New creates a new Exporter.
 func New(ctx context.Context, wg *sync.WaitGroup, store *metrics.Store, options ...Option) (*Exporter, error) {
 	if store == nil {
-		return nil, errors.New("exporter needs a Store")
+		return nil, ErrNeedsStore
 	}
 	if wg == nil {
-		return nil, errors.New("exporter needs a WaitGroup")
+		return nil, ErrNeedsWaitgroup
 	}
 	e := &Exporter{
 		ctx:      ctx,

--- a/internal/exporter/export_test.go
+++ b/internal/exporter/export_test.go
@@ -23,17 +23,10 @@ const prefix = "prefix"
 func TestCreateExporter(t *testing.T) {
 	ctx, cancel := context.WithCancel(context.Background())
 	var wg sync.WaitGroup
-	_, err := New(ctx, &wg, nil)
-	if err == nil {
-		t.Error("expecting error, got nil")
-	}
-	cancel()
-	wg.Wait()
-	ctx, cancel = context.WithCancel(context.Background())
 	store := metrics.NewStore()
-	_, err = New(ctx, &wg, store)
+	_, err := New(ctx, &wg, store)
 	if err != nil {
-		t.Errorf("unexpected error:%s", err)
+		t.Errorf("New(ctx, wg, store) unexpected error: %v", err)
 	}
 	cancel()
 	wg.Wait()
@@ -47,6 +40,20 @@ func TestCreateExporter(t *testing.T) {
 	}
 	cancel()
 	wg.Wait()
+}
+
+func TestNewErrors(t *testing.T) {
+	ctx := context.Background()
+	store := metrics.NewStore()
+	var wg sync.WaitGroup
+	_, err := New(ctx, nil, store)
+	if err == nil {
+		t.Error("New(ctx, nil, store) expecting error, received nil")
+	}
+	_, err = New(ctx, &wg, nil)
+	if err == nil {
+		t.Error("New(ctx, wg, nil) expecting error, received nil")
+	}
 }
 
 func FakeSocketWrite(f formatter, m *metrics.Metric) []string {

--- a/internal/mtail/exec_integration_test.go
+++ b/internal/mtail/exec_integration_test.go
@@ -6,11 +6,12 @@ package mtail_test
 import (
 	"context"
 	"errors"
-	"log"
 	"os/exec"
 	"path/filepath"
 	"testing"
 	"time"
+
+	"github.com/golang/glog"
 )
 
 var mtailPath string
@@ -21,7 +22,7 @@ func init() {
 		err = nil
 	}
 	if err != nil {
-		log.Fatal(err)
+		glog.Infof("exec_integration_test init(): %v", err)
 	}
 	mtailPath = path
 }

--- a/internal/mtail/exec_integration_test.go
+++ b/internal/mtail/exec_integration_test.go
@@ -31,7 +31,9 @@ func TestExecMtail(t *testing.T) {
 		t.Skip()
 	}
 
-	cs := []string{"-progs", "../../examples",
+	cs := []string{
+		"-progs",
+		"../../examples",
 		"-logs", "testdata/rsyncd.log",
 		"-one_shot",
 		"-one_shot_format=prometheus",

--- a/internal/mtail/exec_integration_test.go
+++ b/internal/mtail/exec_integration_test.go
@@ -1,0 +1,46 @@
+// Copyright 2024 Google Inc.  ll Rights Reserved.
+// This file is available under the Apache license.
+
+package mtail_test
+
+import (
+	"context"
+	"errors"
+	"log"
+	"os/exec"
+	"testing"
+	"time"
+)
+
+var mtailPath string
+
+func init() {
+	path, err := exec.LookPath("../../mtail")
+	if errors.Is(err, exec.ErrDot) {
+		err = nil
+	}
+	if err != nil {
+		log.Fatal(err)
+	}
+	mtailPath = path
+}
+
+func TestExecMtail(t *testing.T) {
+	if mtailPath == "" {
+		t.Log("mtail binary not found, skipping")
+		t.Skip()
+	}
+
+	cs := []string{"-progs", "../../examples",
+		"-logs", "testdata/rsyncd.log",
+		"-one_shot",
+		"-one_shot_format=prometheus",
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+	defer cancel()
+	cmd := exec.CommandContext(ctx, mtailPath, cs...)
+	if out, err := cmd.CombinedOutput(); err != nil {
+		t.Logf("%s", out)
+		t.Error(err)
+	}
+}

--- a/internal/mtail/exec_integration_test.go
+++ b/internal/mtail/exec_integration_test.go
@@ -8,6 +8,7 @@ import (
 	"errors"
 	"log"
 	"os/exec"
+	"path/filepath"
 	"testing"
 	"time"
 )
@@ -15,7 +16,7 @@ import (
 var mtailPath string
 
 func init() {
-	path, err := exec.LookPath("../../mtail")
+	path, err := exec.LookPath(filepath.Join("..", "..", "mtail"))
 	if errors.Is(err, exec.ErrDot) {
 		err = nil
 	}

--- a/internal/runtime/runtime.go
+++ b/internal/runtime/runtime.go
@@ -246,6 +246,9 @@ func New(lines <-chan *logline.LogLine, wg *sync.WaitGroup, programPath string, 
 	if store == nil {
 		return nil, errors.New("loader needs a store")
 	}
+	if wg == nil {
+		return nil, errors.New("loader needs a WaitGroup")
+	}
 	r := &Runtime{
 		ms:            store,
 		programPath:   programPath,

--- a/internal/runtime/runtime.go
+++ b/internal/runtime/runtime.go
@@ -241,13 +241,18 @@ type Runtime struct {
 	signalQuit chan struct{} // When closed stops the signal handler goroutine.
 }
 
+var (
+	ErrNeedsStore     = errors.New("loader needs a store")
+	ErrNeedsWaitgroup = errors.New("loader needs a WaitGroup")
+)
+
 // New creates a new program loader that reads programs from programPath.
 func New(lines <-chan *logline.LogLine, wg *sync.WaitGroup, programPath string, store *metrics.Store, options ...Option) (*Runtime, error) {
 	if store == nil {
-		return nil, errors.New("loader needs a store")
+		return nil, ErrNeedsStore
 	}
 	if wg == nil {
-		return nil, errors.New("loader needs a WaitGroup")
+		return nil, ErrNeedsWaitgroup
 	}
 	r := &Runtime{
 		ms:            store,

--- a/internal/runtime/runtime_test.go
+++ b/internal/runtime/runtime_test.go
@@ -25,6 +25,20 @@ func TestNewRuntime(t *testing.T) {
 	wg.Wait()
 }
 
+func TestNewRuntimeErrors(t *testing.T) {
+	store := metrics.NewStore()
+	lines := make(chan *logline.LogLine)
+	var wg sync.WaitGroup
+	_, err := New(lines, nil, "", store)
+	if err == nil {
+		t.Error("New(..., nil) expecting error, got nil")
+	}
+	_, err = New(lines, &wg, "", nil)
+	if err == nil {
+		t.Error("New(..., nil) expecting error, got nil")
+	}
+}
+
 func TestCompileAndRun(t *testing.T) {
 	testProgram := "/$/ {}\n"
 	store := metrics.NewStore()

--- a/internal/tailer/logstream/logstream.go
+++ b/internal/tailer/logstream/logstream.go
@@ -54,6 +54,9 @@ var (
 // channel.  `seekToStart` is only used for testing and only works for regular
 // files that can be seeked.
 func New(ctx context.Context, wg *sync.WaitGroup, waker waker.Waker, pathname string, lines chan<- *logline.LogLine, oneShot bool) (LogStream, error) {
+	if wg == nil {
+		return nil, errors.New("logstream needs a WaitGroup")
+	}
 	u, err := url.Parse(pathname)
 	if err != nil {
 		return nil, err

--- a/internal/tailer/logstream/logstream.go
+++ b/internal/tailer/logstream/logstream.go
@@ -46,6 +46,7 @@ var (
 	ErrUnsupportedURLScheme = errors.New("unsupported URL scheme")
 	ErrUnsupportedFileType  = errors.New("unsupported file type")
 	ErrEmptySocketAddress   = errors.New("socket address cannot be empty, please provide a unix domain socket filename or host:port")
+	ErrNeedsWaitgroup       = errors.New("logstream needs a waitgroup")
 )
 
 // New creates a LogStream from the file object located at the absolute path
@@ -55,7 +56,7 @@ var (
 // files that can be seeked.
 func New(ctx context.Context, wg *sync.WaitGroup, waker waker.Waker, pathname string, lines chan<- *logline.LogLine, oneShot bool) (LogStream, error) {
 	if wg == nil {
-		return nil, errors.New("logstream needs a WaitGroup")
+		return nil, ErrNeedsWaitgroup
 	}
 	u, err := url.Parse(pathname)
 	if err != nil {

--- a/internal/tailer/logstream/logstream_test.go
+++ b/internal/tailer/logstream/logstream_test.go
@@ -2,6 +2,7 @@ package logstream
 
 import (
 	"context"
+	"sync"
 	"testing"
 )
 
@@ -10,5 +11,10 @@ func TestNewErrors(t *testing.T) {
 	_, err := New(ctx, nil, nil, "", nil, false)
 	if err == nil {
 		t.Errorf("New(ctx, nil) expecting error, received nil")
+	}
+	var wg sync.WaitGroup
+	_, err = New(ctx, &wg, nil, ":a/b", nil, false)
+	if err == nil {
+		t.Error("New(ctg, wg, ..., path) expecting error, received nil")
 	}
 }

--- a/internal/tailer/logstream/logstream_test.go
+++ b/internal/tailer/logstream/logstream_test.go
@@ -1,0 +1,14 @@
+package logstream
+
+import (
+	"context"
+	"testing"
+)
+
+func TestNewErrors(t *testing.T) {
+	ctx := context.Background()
+	_, err := New(ctx, nil, nil, "", nil, false)
+	if err == nil {
+		t.Errorf("New(ctx, nil) expecting error, received nil")
+	}
+}

--- a/internal/tailer/tail.go
+++ b/internal/tailer/tail.go
@@ -127,7 +127,10 @@ func (opt logstreamPollWaker) apply(t *Tailer) error {
 	return nil
 }
 
-var ErrNoLinesChannel = errors.New("Tailer needs a lines channel")
+var (
+	ErrNoLinesChannel = errors.New("Tailer needs a lines channel")
+	ErrNeedsWaitgroup = errors.New("tailer needs a WaitGroup")
+)
 
 // New creates a new Tailer.
 func New(ctx context.Context, wg *sync.WaitGroup, lines chan<- *logline.LogLine, options ...Option) (*Tailer, error) {
@@ -135,7 +138,7 @@ func New(ctx context.Context, wg *sync.WaitGroup, lines chan<- *logline.LogLine,
 		return nil, ErrNoLinesChannel
 	}
 	if wg == nil {
-		return nil, errors.New("tailer needs a WaitGroup")
+		return nil, ErrNeedsWaitgroup
 	}
 	t := &Tailer{
 		ctx:          ctx,

--- a/internal/tailer/tail.go
+++ b/internal/tailer/tail.go
@@ -134,6 +134,9 @@ func New(ctx context.Context, wg *sync.WaitGroup, lines chan<- *logline.LogLine,
 	if lines == nil {
 		return nil, ErrNoLinesChannel
 	}
+	if wg == nil {
+		return nil, errors.New("tailer needs a WaitGroup")
+	}
 	t := &Tailer{
 		ctx:          ctx,
 		lines:        lines,

--- a/internal/tailer/tail_test.go
+++ b/internal/tailer/tail_test.go
@@ -47,6 +47,21 @@ func TestTail(t *testing.T) {
 	stop()
 }
 
+func TestTailErrors(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	lines := make(chan *logline.LogLine)
+	var wg sync.WaitGroup
+	_, err := New(ctx, nil, lines)
+	if err == nil {
+		t.Error("New(ctx, nil, lines) expecting error, received nil")
+	}
+	_, err = New(ctx, &wg, nil)
+	if err == nil {
+		t.Error("New(ctx, wg, nil) expecting error, received nil")
+	}
+}
+
 func TestHandleLogUpdate(t *testing.T) {
 	ta, lines, awaken, dir, stop := makeTestTail(t)
 

